internal/configs/provider.go

```diff
@@ -290,3 +333,89 @@ func checkProviderNameNormalized(name string, declrange hcl.Range) hcl.Diagnosti
 	}
 	return diags
 }
+
+func (p *Provider) generateForEachProviders(eval *StaticEvaluator) ([]*Provider, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+	if eval == nil {
+		return nil, diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Iteration not allowed in test files",
+			Detail:   "for_each was declared as an provider attribute in a test file",
+			Subject:  p.AliasExpr.Range().Ptr(),
+		})
+	}
+
+	foeachRefsFunc := func(refs []*addrs.Reference) (*hcl.EvalContext, tfdiags.Diagnostics) {
+		var diags tfdiags.Diagnostics
+		evalContext, evalDiags := eval.EvalContext(StaticIdentifier{
+			Module:    eval.call.addr,
+			Subject:   fmt.Sprintf("provider.%s.for_each", p.Name),
+			DeclRange: p.ForEach.Range(),
+		}, refs)
+		return evalContext, diags.Append(evalDiags)
```

Comment:
Seems like we will have duplicated diags if we append them here and in `EvaluateForEachExpression`. The same for count

Reply:
this func is called in EvaluateForEachExpression. So EvaluateForEachExpression returns this diags as it's own.

Reply:
My bad, I thought `diags` variable is from the outer scope. Thanks!